### PR TITLE
feat(cb2-8166): add promotion functionality to v3

### PR DESCRIPTION
**Pull request by naathanbrown**
## Description

Create a promote end point. Also a minor name refactor on TechRecordGet

Related issue: [8166](https://dvsa.atlassian.net/browse/CB2-8166)

## Type of change

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] Feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)

## Checklist:

- [x] I have performed a self-review of my own code
- [x] I have added tests that prove my fix is effective or that my feature works

### CODE REVIEW

File: src/handler/promote.ts [hunk 1]
```diff
@@ -0,0 +1,97 @@
...
+  logger.info('Promote end point called');
+
+  const promoteErrors = validatePromoteErrors(event);
+
+  if (promoteErrors) {
+    return addHttpHeaders(promoteErrors);
+  }
+
+  const systemNumber = decodeURIComponent(event?.pathParameters?.systemNumber as string ?? '');
+  const createdTimestamp = decodeURIComponent(event?.pathParameters?.createdTimestamp as string ?? '');
```
Comment (gjulien-bjss) [inline src/handler/promote.ts:23]:
```suggestion
  const createdTimestamp = decodeURIComponent(event?.pathParameters?.createdTimestamp ?? '');
```
just out of curiosity, does this not work?

File: src/services/database.ts [hunk 2]
```diff
@@ -144,7 +145,7 @@ export const postTechRecord = async (request: TechrecordGet): Promise <Techrecor
     throw new Error('database client failed getting data');
   }
 };
-export const archiveOldCreateCurrentRecord = async (recordToArchive: TechrecordGet, recordToCreate: TechrecordGet): Promise<undefined | Error> => {
+export const archiveOldCreateCurrentRecord = async (recordToArchive: TechRecordGet, recordToCreate: TechRecordGet, secondaryRecordToArchive?: TechRecordGet): Promise<undefined | Error> => {
```
Comment (gjulien-bjss) [inline src/services/database.ts:148]:
```suggestion
export const archiveOldCreateCurrentRecord = async (recordToArchive: TechRecordGet, recordToCreate: TechRecordGet, secondaryRecordToArchive?: TechRecordGet): Promise<undefined | Error> => {
```

File: src/services/database.ts [hunk 3]
```diff
@@ -165,6 +166,16 @@ export const archiveOldCreateCurrentRecord = async (recordToArchive: TechrecordG
...
+  if (secondaryRecordToArchive) {
+    transactParams.TransactItems?.push({
+      Put: {
+        TableName: tableName,
+        Item: marshall(secondaryRecordToArchive),
+        ConditionExpression: 'attribute_exists(systemNumber) AND attribute_exists(createdTimestamp)',
+      },
+    });
+  }
+
```
Comment (gjulien-bjss) [inline src/services/database.ts:178]:
```ts
export const archiveOldCreateCurrentRecord = async (recordsToArchive: TechRecordGet[], recordToCreate: TechRecordGet): Promise<undefined | Error> => {
  logger.info('Preparing Transact Items');

  const transactParams: TransactWriteItemsInput = {
    TransactItems: [
      {
        Put: {
          TableName: tableName,
          Item: marshall(recordToCreate),
        },
      },
    ],
  }; 
 
 recordsToArchive.forEach(record =>
  transactParams.TransactItems?.push(
    {   
      Put: {
        TableName: tableName,
        Item: marshall(record),
        ConditionExpression: 'attribute_exists(systemNumber) AND attribute_exists(createdTimestamp)',
      },
    }
  )
);
```
You could do something like this.

File: tests/unit/handler/promote.test.ts [hunk 4]
```diff
@@ -0,0 +1,107 @@
...
+    const expectedBody = {
+      techRecord_statusCode: 'current',
+      createdTimestamp: mockedDate.toISOString(),
+      techRecord_createdByName: 'user',
+      techRecord_createdById: '123',
+      techRecord_reasonForCreation: 'Just a test for promoting',
+    };
+
+    expect(archiveOldCreateCurrentRecord).toHaveBeenCalledWith(
+      expect.anything(),
```
Comment (gjulien-bjss) [inline tests/unit/handler/promote.test.ts:59]: `expect.anything()` ? Why ?

File: src/services/database.ts (hunk 3 above)
Comment (naathanbrown) [reply src/services/database.ts:178]: yeah I like this

File: src/handler/promote.ts (hunk 1 above)
Comment (kcarr-bjss) [reply src/handler/promote.ts:23]: I tried this on the archive endpoint. The linter definitely does not like this at all and insisted on the "as string" approach.

File: src/handler/promote.ts [hunk 5]
```diff
@@ -0,0 +1,99 @@
...
+  logger.debug(`result is: ${JSON.stringify(provisionalRecord)}`);
+
+  if (!provisionalRecord || !Object.keys(provisionalRecord).length) {
+    return addHttpHeaders({
+      statusCode: 404,
+      body: `No record found matching systemNumber ${systemNumber} and timestamp ${createdTimestamp}`,
+    });
+  }
+
+  if (provisionalRecord.techRecord_statusCode !== 'provisional') {
```
Comment (gjulien-bjss) [inline src/handler/promote.ts:40]: Probably best to have an enum for statusCodes

File: src/handler/promote.ts [hunk 6]
```diff
@@ -0,0 +1,99 @@
...
+  logger.debug(`Current result ${JSON.stringify(currentResult)}`);
+
+  const recordsToArchive: TechRecordGet[] = [];
+
+  if (currentResult.length > 0) {
+    const currentRecord = await getBySystemNumberAndCreatedTimestamp(currentResult[0].systemNumber, currentResult[0].createdTimestamp) as TechRecordGet;
+    currentRecord.techRecord_statusCode = 'archived';
+    currentRecord.techRecord_lastUpdatedAt = new Date().toISOString();
+    currentRecord.techRecord_lastUpdatedByName = userDetails.username;
+    currentRecord.techRecord_lastUpdatedById = userDetails.msOid;
```
Comment (gjulien-bjss) [inline src/handler/promote.ts:59]: I think @shivangidas's PR has some helped functions for the lastUpdatedAt etc... logic. I don't really know how to go implementing them in this PR as well but would be good to standardise the audit logic

Comment (naathanbrown) [reply src/handler/promote.ts:59]: Yeah I've seen that... I think we put hers in first then I just redo a load of stuff tbh

**Review by kcarr-bjss — APPROVED** [review]

File: src/handler/promote.ts (hunk 5 above)
Comment (naathanbrown) [reply src/handler/promote.ts:40]: Where should this go? Does the types package include enums or just declare locally? What do you think?

Comment (shivangidas) [reply src/handler/promote.ts:40]: I was doing them locally

File: src/handler/promote.ts [hunk 7]
```diff
@@ -0,0 +1,100 @@
...
+
+  const promoteErrors = validatePromoteErrors(event);
+
+  if (promoteErrors) {
+    return addHttpHeaders(promoteErrors);
+  }
+
+  const systemNumber = decodeURIComponent(event?.pathParameters?.systemNumber ?? '');
+  const createdTimestamp = decodeURIComponent(event?.pathParameters?.createdTimestamp ?? '');
+  const userDetails = getUserDetails(event.headers.Authorization ?? '');
```
Comment (shivangidas) [inline src/handler/promote.ts:25]: getUserDetail needs to be in the try block, it'll throw and error if name and id is missing

Comment (naathanbrown) [reply src/handler/promote.ts:25]: tbh all of it needs to be? Really unsure why the try is where it is

**Review by shivangidas — APPROVED** [review]

**Pull request merged**